chore: add rotating log for audit data

DmitriyMV opened the pull request:
Adds rotating audit log writer. Also minor improvements.

For #37

Unix4ever commented on internal/backend/server.go:
```diff
@@ -352,13 +352,15 @@ func (s *Server) buildServerOptions() ([]grpc.ServerOption, error) {
 
-	unaryInterceptors = append(unaryInterceptors, unaryAuthInterceptors...)
-	streamInterceptors = append(streamInterceptors, streamAuthInterceptors...)
+	for _, authIntereceptor := range authIntereceptors {
```
```suggestion
	for _, authInterceptor := range authInterceptors {
```

DmitriyMV replied:
Thx!

smira commented on internal/backend/runtime/omni/audit/log_file.go:
```diff
@@ -0,0 +1,87 @@
+	return l.writeAt(data, time.Now())
+}
+
+func (l *LogFile) writeAt(data any, at time.Time) error {
```
I would rename this to something else to avoid confusion with `io.WriterAt`

DmitriyMV replied:
Renamed to `dumpAt` and `Write` -> `Dump`.

smira commented on internal/backend/runtime/omni/audit/log_file.go:
```diff
@@ -0,0 +1,87 @@
+	// into the current day (happens on day's border). One solution is to have two decriptors opened:
+	// one for the current day and one for the previous day. We can close the previous day's descriptor
+	// when we open the 3rd day's descriptor. This way we will have at most two descriptors opened.
+	if l.f != nil && ordered.MakeTriple(at.Date()).Compare(ordered.MakeTriple(l.lastWrite.Date())) <= 0 {
```
idk, but it feels it might be easier to run a goroutine which "flips" the file at midnight, and all writes go to a single currently opn file, so they are at least chronologically ordered?

DmitriyMV replied:
I reworked this entirely, more explanation below.

smira commented on internal/backend/runtime/omni/audit/log_file.go:
```diff
@@ -0,0 +1,87 @@
+		return err
+	}
+
+	if err := json.NewEncoder(f).Encode(data); err != nil {
```
this won't work well - we use direct file (without buffering), and json encoder would issue tons of small writes directly to the OS.

plus this won't be atomic, as append mode is only guaranteed for a single write (records will be interleaved)

in this case I would say it's easier to to write to the `bytes.Buffer` and then write bytes as a single write to the file.

DmitriyMV replied:
Yeah, I agree.

DmitriyMV commented:
So on further thinking I reworked `writeAt` (now `dumpAt` because of the similarity to `spew.Dump` and other pretty printers/dumpers) to take the lock right after we are finished with encoding. I looked at other solutions (such as creating locked wrapper around the file and implement synced `WriteTo`), but overall I don't think it worth it. I also moved the time assertion inside the locked part, so now all writers are fully ordered.

DmitriyMV commented:
@smira I also added `TestLogFile_CurrentDayConcurrent` test to be sure.

smira commented on internal/backend/runtime/omni/audit/log_file.go:
```diff
@@ -0,0 +1,107 @@
+
+	logPath := filepath.Join(l.dir, at.Format("2006-01-02")) + ".jsonlog"
+
+	f, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
```
not exactly clear how this could would recover from a failure.

I would say `.Close()` error should be ignored (as there's no recovery path, and it doesn't actually fail).

While we should zero `l.f` to make the code jump into creating a new file.

Creating new file might fail vs. closing it

DmitriyMV replied:
> not exactly clear how this could would recover from a failure.

I'm not sure it can without alerting the user. We can create a temporary ring buffer if we failed to create a file, but then what? Try again at the next write (and the next after that and so on) and write all stuff which we stored if we succeed?

My current idea is to write metrics/logs if the error occured.

> While we should zero l.f to make the code jump into creating a new file.

Yeah, agreed.

smira replied:
if we failed to create the file, the audit log will have missing entries, but once the error condition is clear, it should continue to log audit entries

smira commented on internal/backend/runtime/omni/audit/log_file.go:
```diff
@@ -0,0 +1,107 @@
+		return err
+	}
+
+	_, err = io.Copy(f, b)
```
if we just `Write(b.Bytes())` that would be atomic, we don't need to lock the writes.

DmitriyMV replied:
Yeah, but that will break `lastWrite` field access, and since `*os.File.Write` uses mutex internally I don't think it makes a lot of sense to unlock and then atomically write `lastWrite` field.

DmitriyMV replied:
Replaced `io.Copy` with `b.WriteTo(f)`

DmitriyMV commented:
/m
